dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs

```diff
@@ -0,0 +1,473 @@
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task UpsertAsync(string collectionName,
+        string key,
+        string? metadata,
+        string embedding,
+        DateTimeOffset? timestamp,
+        CancellationToken cancellationToken = default)
+    {
+        collectionName = NormalizeIndexName(collectionName);
+
+        using SqlCommand cmd = this._connection.CreateCommand();
+
+        cmd.CommandText = $@"
+                MERGE INTO {this.GetFullTableName(this._configuration.MemoryTableName)}
+                USING (SELECT @key) as [src]([key])
+                ON {this.GetFullTableName(this._configuration.MemoryTableName)}.[key] = [src].[key]
+                WHEN MATCHED THEN
+                    UPDATE SET metadata=@metadata, embedding=@embedding, timestamp=@timestamp
+                WHEN NOT MATCHED THEN
+                    INSERT ([id], [collection], [key], [metadata], [timestamp], [embedding])
+                    VALUES (NEWID(), @collection, @key, @metadata, @timestamp, @embedding);
+
+
+                MERGE {this.GetFullTableName($"{this._configuration.EmbeddingsTableName}_{collectionName}")} AS [tgt]
+                USING (
+                    SELECT
+                        {this.GetFullTableName(this._configuration.MemoryTableName)}.[id],
+                        cast([vector].[key] AS INT) AS [vector_value_id],
+                        cast([vector].[value] AS FLOAT) AS [vector_value]
+                    FROM {this.GetFullTableName(this._configuration.MemoryTableName)}
+                    CROSS APPLY
+                        openjson(@embedding) [vector]
+                    WHERE {this.GetFullTableName(this._configuration.MemoryTableName)}.[key] = @key
+                        AND {this.GetFullTableName(this._configuration.MemoryTableName)}.[collection] = @collection
+                ) AS [src]
+                ON [tgt].[memory_id] = [src].[id] AND [tgt].[vector_value_id] = [src].[vector_value_id]
+                WHEN MATCHED THEN
+                    UPDATE SET [tgt].[vector_value] = [src].[vector_value]
+                WHEN NOT MATCHED THEN
+                    INSERT ([memory_id], [vector_value_id], [vector_value])
+                    VALUES ([src].[id],
+                            [src].[vector_value_id],
+                            [src].[vector_value] )
+                ;";
+
+        cmd.Parameters.AddWithValue("@collection", collectionName);
+        cmd.Parameters.AddWithValue("@key", key);
+        cmd.Parameters.AddWithValue("@metadata", metadata ?? (object)DBNull.Value);
+        cmd.Parameters.AddWithValue("@embedding", embedding);
+        cmd.Parameters.AddWithValue("@timestamp", timestamp ?? (object)DBNull.Value);
+
+        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    private string GetFullTableName(string tableName)
+    {
+        return $"[{this._configuration.Schema}].[{tableName}]";
```

Comment:
Proper escaping. I.e. what if the `tableName` is `foo]bar` (yes, that's stupid, but still).

Reply:
Added escaping some characters in SLQL object names.